storage/remote/read.go

```diff
@@ -79,40 +54,124 @@ func (q *querier) Select(matchers ...*labels.Matcher) storage.SeriesSet {
 func (q *querier) Close() error {
 	return nil
 }
 
+// ExternablLabelsHandler returns a storage.Queryable which creates a
+// externalLabelsQuerier.
+func ExternablLabelsHandler(next storage.Queryable, externalLabels model.LabelSet) storage.Queryable {
+	return storage.QuerierFunc(func(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+		q, err := next.Querier(ctx, mint, maxt)
+		if err != nil {
+			return nil, err
+		}
+		return &externalLabelsQuerier{Querier: q, externalLabels: externalLabels}, nil
+	})
+}
+
+// externalLabelsQuerier is a querier which ensures that Select() results match
+// the configured external labels.
+type externalLabelsQuerier struct {
+	storage.Querier
+
+	externalLabels model.LabelSet
+}
+
+// Select adds equality matchers for all external labels to the list of matchers
+// before calling the wrapped storage.Queryable. The added external labels are
+// removed from the returned series sets.
+func (q externalLabelsQuerier) Select(matchers ...*labels.Matcher) storage.SeriesSet {
+	m, added := addExternalLabels(q.externalLabels, matchers)
+	s := q.Select(m...)
+	return newSeriesSetFilter(s, added)
+}
+
+// PreferLocalStorageFilter returns a QuerierFunc which creates a NoopQuerier if
+// requested timeframe can be answered completely by the local TSDB, and reduces
+// maxt if the timeframe can be partially answered by TSDB.
+func PreferLocalStorageFilter(next storage.Queryable, cb startTimeCallback) storage.Queryable {
+	return storage.QuerierFunc(func(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+		localStartTime, err := cb()
+		if err != nil {
+			return nil, err
+		}
+		cmaxt := maxt
+		// Avoid queries whose timerange is later than the first timestamp in local DB.
+		if mint > localStartTime {
+			return storage.NoopQuerier(), nil
+		}
+		// Query only samples older than the first timestamp in local DB.
+		if maxt > localStartTime {
+			cmaxt = localStartTime
+		}
+		return next.Querier(ctx, mint, cmaxt)
+	})
+}
+
+// RequiredLabelsFilter returns a storage.Queryable which creates a
+// requiredLabelsQuerier.
+func RequiredLabelsFilter(next storage.Queryable, requiredLabels model.LabelSet) storage.Queryable {
+	return storage.QuerierFunc(func(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+		q, err := next.Querier(ctx, mint, maxt)
+		if err != nil {
+			return nil, err
+		}
+		return &requiredLabelsQuerier{Querier: q, requiredLabels: requiredLabels}, nil
+	})
+}
+
+// requiredLabelsQuerier wraps a storage.Querier and requires Select() calls to
+// match the given labelSet.
+type requiredLabelsQuerier struct {
+	storage.Querier
+
+	requiredLabels model.LabelSet
+}
+
+// Select returns a NoopSeriesSet if the given matchers don't match the label
+// set of the requiredLabelsQuerier. Otherwise it'll call the wrapped querier.
+func (q requiredLabelsQuerier) Select(matchers ...*labels.Matcher) storage.SeriesSet {
+	ls := q.requiredLabels
```

Comment:
Is this a copy or a reference?

Reply:
Good catch! This function has now side effects after I changed the labels type from a slice labels.Labels to the map mode.LabelSet. Fixing